Add ID for notifications

MatthewPhinney opened the pull request.

MatthewPhinney commented:
Resolve #80. Resolve #83.

clark800 commented on test/putTransferSpec.js:
```diff
@@ -974,9 +974,11 @@ describe('PUT /transfers/:id', function () {
     const notification = nock('http://subscriber.example')
       .post('/notifications', (body) => {
+        const idParts = body.id.split('/')
+        const notificationId = body.id.split('/')[idParts.length - 1]
```
Looks like you can use the `idParts` here to avoid splitting a second time

MatthewPhinney replied:
Oops, meant to fix that earlier ....

clark800 commented:
Can you add a test case?

clark800 commented:
Is it possible to GET an individual notification by id?

MatthewPhinney commented:
As far as I know, we do not support that. @justmoon do we intend to support querying notifications by ID?

justmoon commented:
> @justmoon do we intend to support querying notifications by ID?

Yes, regarding the authentication of certificates, I would propose that we give people two options:

1. Validate the TLS client certificate when the notification is submitted. - Widely supported, standards-based solution.

2. GET the notification from the origin server. - Some developer may have difficulty with method 1. Perhaps their TLS implementation doesn't support validating client certs very well. So as an alternative, they can GET the notification from the origin server simply by doing `fetch(notification.id).then(body => if (deepEquals(notification, body)) { ... })`.

Pattern 2 is less efficient, but extremely simple. One implementation note: User should GET the notification (and process it) **before** responding positively to the PUT request. After the receiving server has acknowledged the PUT, the sending server (in this case the ledger) should be free to delete their record of the notification if they want.

MatthewPhinney commented:
@clark800 what kind of tests did you have in mind? The PR only changes the format of the notification, and I updated the existing test to expect this new format.

clark800 commented:
I was looking for a test of GET, then I noticed that it wasn't implemented

MatthewPhinney commented:
@clark800 can you please review this?

clark800 commented:
What about the GET endpoint?

MatthewPhinney commented:
I can add that, but I don't see any information about it in the design docs. Maybe we could track it with a separate JIRA issue that has more design details?

clark800 commented:
It's just the REST model, if you have a URI it should actually "exist" in the sense of having some kind of behavior at that URI.

emschwartz commented:
+1 the whole point of adding IDs is to allow clients to `GET` them from the ledger.

See Stefan's comment from above:
> GET the notification from the origin server. - Some developer may have difficulty with method 1. Perhaps their TLS implementation doesn't support validating client certs very well. So as an alternative, they can GET the notification from the origin server simply by doing fetch(notification.id).then(body => if (deepEquals(notification, body)) { ... }).

MatthewPhinney commented:
That makes sense, I was just thinking we could have a design discussion elsewhere. But we may as well do it here. Should the response to GET `/subscriptions/:subscription_id/notifications/:notification_id` be the same as the payload of the POST notification that the ledger performs with queued notifications?
```
{
  "id": "http://eur-ledger.example/subscriptions/52a42d6f-8d9c-4c05-b31c-cccc8bbdb30d/notifications/fccacc64-8b8d-4dd6-91f5-8ffc7674e8be",
  "subscription": "http://eur-ledger.example/subscriptions/52a42d6f-8d9c-4c05-b31c-cccc8bbdb30d",
  "event": "transfer.update",
  "resource": {
    // ...
  }
}
```
I've implemented most of this but want to make sure we're all on the same page before flushing out the code and adding proper tests.

clark800 commented:
> Should the response to GET /subscriptions/:subscription_id/notifications/:notification_id be the same as the payload of the POST notification that the ledger performs with queued notifications?

Is there a reason that it should be different?

MatthewPhinney commented on src/models/notifications.js:
```diff
@@ -0,0 +1,43 @@
+    include: [{ all: true }]})
+  if (!notification) {
+    throw new NotFoundError('Unknown notification ID')
+  } else if (notification.Subscription.id !== subscriptionId) {
```
This will throw if `notification.Subscription` is `undefined`, which should never happen, but is possible if the database becomes corrupted. I can add a check for this case, if we think it's worthwhile.

clark800 replied:
Seems worthwhile

MatthewPhinney commented on src/models/notifications.js:
```diff
@@ -0,0 +1,43 @@
+const Notification = require('./db/notification').Notification
+const Transfer = require('./db/transfer').Transfer
+
+function isOwnerOrAdmin (requestingUser, subscription) {
```
This method is taken from the `subscriptions` model. It may make sense to put it in a separate file of library methods that both the `notifications` model and the `subscriptions` model can access, to avoid code duplication.

clark800 replied:
I'm all for reducing duplication

MatthewPhinney commented:
The GET notifications endpoint has been implemented.

clark800 commented on test/data/notificationExample.json:
```diff
@@ -0,0 +1,5 @@
+{
```
I think this is a database entry right? At first I interpreted this as a notification response. Maybe rename to something like "notificationDatabaseEntry.json"?

MatthewPhinney commented:
I moved the subscription util methods into a separate file, so that we can use them in both the `notifications` model and the `subscriptions` model. I also added a check to ensure the notification has a corresponding subscription in the DB. I did not add a test for this specific case, since doing so requires adding an example notification which has a subscription ID for a non-existent subscription. Doing this causes a FK constraint error.

clark800 commented on src/models/notifications.js:
```diff
@@ -0,0 +1,41 @@
+    include: [{ all: true }]})
+  if (!notification) {
+    throw new NotFoundError('Unknown notification ID')
+  } else if (notification.Subscription === undefined) {
```
I'm not sure how sequelize works here, would it be `undefined` or `null`?

MatthewPhinney replied:
Good point, I checked and it's `null`. I updated the code to reflect this.

clark800 commented:
LGTM
